# 英国EC企業の販売データの評価とクリーニング

## 分析目標

このデータ分析の目的は、市場の販売データに基づいて売れ筋商品を掘り起こし、売上向上につながるより効果的なマーケティング戦略を策定することである。
本実践プロジェクトの目的は、データのクリーンさと整然性を評価する練習を行い、その評価結果に基づいてデータをクリーニングすることで、次の分析に利用できるデータを得ることである。

## プロジェクト概要

元のデータセットには、英国のオンライン小売企業における2010年12月1日から2011年12月9日までのすべての取引記録が含まれており、世界各国・地域における同社の事業データを網羅している。同社は主に、誕生日プレゼント、結婚記念品、クリスマスギフトなど、さまざまな場面に対応したギフト商品を販売している。顧客層は主に卸売業者と個人消費者で構成されており、その中でも卸売業者がかなり大きな割合を占めている。

データの各列の意味は以下のとおりである。

* `InvoiceNo`：請求書番号。6桁の数字で、取引の一意の識別子として使用される。このコードが文字「C」で始まる場合、その取引がキャンセルされたことを示す。
* `StockCode`：商品コード。5桁の数字で、商品の一意の識別子として使用される。
* `Description`：商品名。
* `Quantity`：取引における商品の数量。
* `InvoiceDate`：請求書の日付と時刻。取引が発生した日時を表す。
* `UnitPrice`：単価。価格の単位は英ポンド（£）である。
* `CustomerID`：顧客番号。5桁の数字で、顧客の一意の識別子として使用される。
* `Country`：国名。顧客が居住している国の名称を表す。


## データの読み込み

データ分析に必要なライブラリをインポートし、Pandasの`read_csv`関数を使用して、元データファイル`e_commerce.csv`のデータ内容をDataFrameとして読み込み、変数`original_data`に代入する。

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## データの評価

この部分では、前の部分で作成した`original_data`というDataFrameに含まれるデータを評価する。

評価は主に、構造と内容という2つの観点から行う。つまり、データの整然性とクリーンさを確認する。データの構造的な問題とは、「各列が1つの変数を表し、各行が1つの観測値を表し、各セルが1つの値を表す」という3つの基準を満たしていない状態を指す。データの内容上の問題には、欠損データ、重複データ、無効なデータなどが含まれる。

### データの整然性の評価

In [4]:
original_data.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
250465,559032,22382,LUNCH BAG SPACEBOY DESIGN,100,7/5/2011 14:53,1.45,17511.0,United Kingdom
8510,537137,21421,PORCELAIN ROSE LARGE,3,12/5/2010 12:43,1.25,16327.0,United Kingdom
348014,567369,21415,CLAM SHELL SMALL,6,9/20/2011 9:06,2.10,17527.0,United Kingdom
472446,576688,23083,SET 6 PAPER TABLE LANTERN STARS,6,11/16/2011 12:16,3.75,16717.0,United Kingdom
353819,567846,15060B,FAIRY CAKE DESIGN UMBRELLA,4,9/22/2011 13:13,3.75,13486.0,United Kingdom
56544,541104,22300,COFFEE MUG DOG + BALL DESIGN,1,1/13/2011 14:29,4.96,NaN,United Kingdom
78427,542881,84884A,ANT WHITE WIRE HEART SPIRAL,2,2/1/2011 12:54,3.95,14587.0,United Kingdom
121701,546778,22423,REGENCY CAKESTAND 3 TIER,2,3/16/2011 16:38,12.75,14572.0,United Kingdom
219101,556083,20717,STRAWBERRY SHOPPER BAG,1,6/8/2011 15:23,2.46,NaN,United Kingdom
291260,562444,21830,ASSORTED CREEPY CRAWLIES,24,8/4/2011 18:44,0.42,14713.0,United Kingdom


抽出した10行のサンプルデータを見る限り、データは「各列が1つの変数を表し、各行が1つの観測値を表し、各セルが1つの値を表す」という基準を満たしている。

具体的には、各行はある商品の1回の取引を表し、各列は取引に関連する各変数を表しているため、構造的な問題は存在しない。

### データのクリーンさの評価

In [5]:
original_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


出力結果を見ると、データには合計541,909件の観測値があり、`Description`と`CustomerID`の変数に欠損値が存在していることが分かる。

また、`InvoiceDate`のデータ型は日付型、`CustomerID`のデータ型は文字列型であるべきであるため、データ型の変換を行う必要がある。

#### 欠損データの評価

`Description`に欠損値が存在することを確認した後、条件を指定して欠損値を含む観測値を抽出する。

In [6]:
original_data[original_data["Description"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


`Description`変数の値が欠損している取引データは1,454件存在する。

出力結果を見ると、これらの`Description`が欠損している取引データでは、`UnitPrice`がすべて0であることが分かる。この推測を検証するために、さらに抽出条件を追加し、`Description`変数が欠損しており、かつ`UnitPrice`が0ではないデータが存在するかどうかを確認する。

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


抽出結果の件数は0件であり、`Description`の値が欠損しているデータは、同時に有効な`UnitPrice`の値も持っていないことが分かる。

`Description`は商品名を表し、`UnitPrice`は商品の単価を表す。いずれも今後の商品取引分析において重要な変数である。これらが同時に欠損または無効である場合、そのデータは有効な意味を提供できないと考えられるため、後続の処理で削除することができる。

`CustomerID`変数にも同様に欠損値が存在するため、条件を指定して欠損値を含む観測値を抽出する。

In [8]:
original_data[original_data["CustomerID"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/2010 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/2011 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/2011 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/2011 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/2011 10:26,10.79,NaN,United Kingdom


`CustomerID`は顧客番号を表す変数であり、売れ筋商品を分析するうえで必須の変数ではない。

また、出力結果を見ると、`CustomerID`が欠損している取引データの中にも有効なデータが存在していることが分かる。そのため、`CustomerID`が空欄の観測値はそのまま残すこととする。

#### 重複データの評価

データ変数の意味から見ると、`InvoiceNo`、`StockCode`、`CustomerID`はいずれも一意の識別子である。

しかし、1回の取引には複数の商品が含まれる可能性があるため、`InvoiceNo`は重複する場合がある。また、異なる取引に同じ商品が含まれる可能性があるため、`StockCode`も重複する場合がある。さらに、顧客は複数回取引を行ったり、複数の商品を注文したりする可能性があるため、`CustomerID`も重複する場合がある。

したがって、このデータセットについては、重複データを評価する必要はない。

#### 不一致データの評価

不一致データは`Country`変数に存在する可能性があるため、同一の国を指す複数の異なる値が存在するかどうかを確認する。

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


`Country`変数の値を見ると、`"USA"`と`"United States"`はいずれもアメリカを表しており、`"United Kingdom"`、`"UK"`、`"U.K."`はいずれもイギリスを表している。

そのため、クリーニングの段階でこれらの値を統一し、1つの表記にそろえる必要がある。

#### 無効または誤ったデータの評価

DataFrameの`describe`メソッドを使用することで、数値データの統計情報を素早く把握することができる。

In [10]:
original_data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


出力結果を見ると、`Quantity`と`UnitPrice`に負の値が存在しており、後続の数値分析に影響を与える可能性がある。

そのため、まず`Quantity`の値が負数である観測値を抽出し、その意味をさらに評価する。

In [11]:
original_data[original_data["Quantity"] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


抽出結果を見ると、`Quantity`変数が負数である観測値では、`InvoiceNo`がいずれもCで始まっているように見える。これは、注文がキャンセルされたことを表していると考えられる。

ただし、この推測が必ずしも正しいとは限らない。そこで、この推測を検証するために、さらに抽出条件を追加し、`Quantity`変数が負数であり、かつ`InvoiceNo`がCで始まらない観測値が存在するかどうかを確認する。

In [12]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/2010 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/2010 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/2010 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/2011 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/2011 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/2011 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/2011 9:56,0.0,NaN,United Kingdom


上記の推測は誤りであることが確認された。なぜなら、`Quantity`変数が負数であり、かつ`InvoiceNo`がCで始まらない観測値も存在しているためである。

ただし、上記の出力結果によると、これらの抽出された観測値の`UnitPrice`はすべて0である。そのため、`UnitPrice`の条件を追加して検証する。

In [13]:
original_data[(original_data["Quantity"] < 0)&(original_data["InvoiceNo"].str[0] != "C")&(original_data["UnitPrice"] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


出力結果により、この推測は検証された。`Quantity`変数が負数である場合、観測値は以下のいずれかの条件を満たしている。

- `InvoiceNo`がCで始まっており、注文がキャンセルされたことを示している
- `UnitPrice`が0であり、単価が0ポンドであることを示している

これらの取引データはいずれも有効な成約データではなく、売上に貢献しないため、後続の分析対象には含めない。

したがって、データクリーニングの段階で、`Quantity`が負数である観測値を削除する。

次に、`UnitPrice`が負数である観測値を抽出する。

In [14]:
original_data[original_data["UnitPrice"] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


出力結果を見ると、`UnitPrice`が負数である観測値はいずれも貸倒損失の調整であり、実際の商品取引データには該当しない。

そのため、データクリーニングの段階で、これらの観測値も削除する。

## データのクリーニング

前の評価部分で得られた結論に基づき、実施する必要のあるデータクリーニングは以下のとおりである。

- `InvoiceDate`変数のデータ型を日時型に変換する
- `CustomerID`変数のデータ型を文字列型に変換する
- `Description`変数が欠損している観測値を削除する
- `Country`変数の値`"USA"`を`"United States"`に置き換える
- `Country`変数の値`"UK"`、`"U.K."`を`"United Kingdom"`に置き換える
- `Quantity`変数の値が負数である観測値を削除する
- `UnitPrice`変数の値が負数である観測値を削除する

クリーニング後のデータと元のデータを区別するために、新しい変数`cleaned_data`を作成し、`original_data`のコピーを代入する。

以降のクリーニング処理は、すべて`cleaned_data`に対して行う。

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


`InvoiceDate`変数のデータ型を日時型に変換する。

In [16]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[us]

`CustomerID`変数のデータ型を文字列型に変換する。

In [17]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: str

`CustomerID`変数の値の末尾にある`".0"`を削除する。

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: str

`Description`変数が欠損している観測値を削除し、削除後にこの列の欠損値の合計数を確認する。

In [19]:
cleaned_data.dropna(subset=["Description"], inplace=True)

In [20]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

`Country`変数の値`"USA"`を`"United States"`に置き換え、置き換え後に`"USA"`という値の件数を確認する。

In [21]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [22]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

`Country`変数の値`"UK"`、`"U.K."`を`"United Kingdom"`に置き換え、置き換え後に`"UK"`と`"U.K."`という値の件数を確認する。

In [23]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})


In [24]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


`Quantity`変数の値が負数である観測値を削除し、削除後に`Quantity`変数の値が負数である件数を確認する。

In [25]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [26]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

`UnitPrice`変数の値が負数である観測値を削除し、削除後に`UnitPrice`変数の値が負数である件数を確認する。

In [27]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [28]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## クリーニング後のデータの保存

データクリーニングが完了した後、クリーンで整然としたデータを新しいファイルに保存する。ファイル名は`e_commerce_cleaned.csv`とする。

In [29]:
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


In [30]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)

In [31]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\tangy\AppData\Local\Temp\ipykernel_16728\2686214642.py:1: DtypeWarning: Columns (0: InvoiceNo) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
